### PR TITLE
Don't use string comparison to determine geometry type

### PR DESCRIPTION
**Pull request by m-kuhn**
Because this kind of title is what you really like as pull request reviewer.

### CODE REVIEW

File: src/core/qgsdataprovider.cpp [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsdataprovider.h"
+
+
+void QgsDataProvider::setProviderProperty( QgsDataProvider::ProviderProperty property, const QVariant& value )
```
Comment (nyalldawson) [inline src/core/qgsdataprovider.cpp]: What do you think about adding unit tests for these?

File: src/core/qgsproject.cpp [hunk 2]
```diff
@@ -2116,6 +2133,25 @@ void QgsProject::snapSettings( QStringList &layerIdList, QStringList &enabledLis
   avoidIntersectionList = readListEntry( "Digitizing", "/AvoidIntersectionsList" );
 }
 
+bool QgsProject::evaluateDefaultValues() const
+{
+  return imp_->evaluateDefaultValues;
+}
+
+void QgsProject::setEvaluateDefaultValues( bool evaluateDefaultValues )
+{
```
Comment (nyalldawson) [inline src/core/qgsproject.cpp]: I'd suggest a unit test for this too

**Comment by nyalldawson** [comment]
These different things look good to me!

File: src/core/qgsproject.cpp (hunk 2 above)
Comment (m-kuhn) [inline src/core/qgsproject.cpp]: I'd suggest this as well

**Comment by m-kuhn** [comment]
At the moment there are two methods to determine the geometry type:

* `QgsWKBTypes::Type wkbType() const`

Which returns the value of an internal `mWkbType` member, controlled by the subclasses of `QgsAbstractGeometryV2`. This is reset to `QgsWKBTypes::Unknown` when `clear()` is called. But the default constructor initializes it to a proper type. Is this on purpose?

Because of this I introduced a new method

    virtual QgsWKBTypes::Type QgsAbstractGeometryV2::flatWkbType() const = 0;

which basically returns static class information and corresponds to the `QString geometryType()` method.

But now I'm becoming unsure if it wouldn't be better to change the behavior of clear().

What do you think @nyalldawson and @mhugent ?

**Comment by mhugent** [comment]
@m-kuhn: yes, I also think it is better to change the behaviour of clear() to set the geometry type to the one in the default constructor (instead to unknown). E.g. an empty linestring is still a linestring.

**Comment by m-kuhn** [comment]
Thanks @mhugent I'll give it a try and see if there are any failing tests.

**Pull request merged**